Create CNIL demo page

nathangavin opened the pull request:
### Description:

Details the CNIL compliance page with a UI demo of the feature. The data used is currently dummy data. The page is only accessible to super users, under Admin -> Privacy -> Compliance.

Page is hidden behind the PrivacyCompliance feature flag.

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

caddoo committed:
Add compliance page UI

caddoo committed:
Update naming

caddoo committed:
Fix linting problems

nathangavin committed:
revert some unnecessary changes

nathangavin committed:
Update text as per feedback and migrate text to lang files

nathangavin committed:
use translate string for save button

nathangavin committed:
replace table column headers with translate string

nathangavin committed:
Merge branch '5.x-dev' into dev-19376-UI

nathangavin committed:
Merge branch '5.x-dev' into dev-19376-UI

nathangavin committed:
Update CNIL text to be more accurate

michalkleiner committed:
Fix vue imports

michalkleiner committed:
Build dist files

michalkleiner commented on plugins/PrivacyManager/lang/en.json:
```diff
@@ -13,6 +13,13 @@
+        "ComplianceTableSettingStatus": "Status",
+        "ComplianceTableSettingNotes": "Notes",
+        "ComplianceSelectSite": "Select a site below to get an indication if the given site is compliant according to the indicated privacy law",
+        "ComplianceEnforceCheckboxIntro": "Enforce Compliance where possible",
```
```suggestion
        "ComplianceEnforceCheckboxIntro": "Enforce compliance where possible",
```

nathangavin replied:
updated

michalkleiner commented on plugins/PrivacyManager/vue/src/Compliance/Compliance.store.ts:
```diff
@@ -0,0 +1,121 @@
+import { DeepReadonly, reactive, readonly } from 'vue';
+
+export interface ComplianceIndicator {
```
Were there possibly other suggested names for this type of object?
For me, an indicator is the green/red status. this feels rather like an aspect of the given compliance, a 'requirement' perhaps?
What do you think about `ComplianceRequirement` or `ComplianceCriterion`?

nathangavin replied:
updated to ComplianceRequirement

michalkleiner commented on plugins/PrivacyManager/vue/src/Compliance/Compliance.vue:
```diff
@@ -9,15 +9,74 @@
+    v-model="site"
+  />
+
+  <div>
```
Do we need the extra wrapping div?

nathangavin replied:
nope, removed

michalkleiner commented on plugins/PrivacyManager/vue/src/Compliance/ComplianceOverview.vue:
```diff
@@ -0,0 +1,101 @@
+      :inline-help="translate('PrivacyManager_ComplianceEnforceCheckboxHelp')"
+      v-model="isComplianceModeEnabled"
+    />
+    <SaveButton
```
The save button could be hidden as well during loading, otherwise it stays visible and can be clicked during loading.

nathangavin replied:
added loading check

michalkleiner commented on plugins/PrivacyManager/vue/src/Compliance/ComplianceTable.vue:
```diff
@@ -0,0 +1,57 @@
+              'icon-close': item.value === 'non_compliant',
+              'icon-circle': item.value === 'unknown'
+           }"></span>
+          {{ item.value }}
```
It might be good to separate the textual/legible value of the compliance requirement since it currently controls both the icon/colour and the visual text, which won't be possible when translatable. Would we expect the compliance requirements to come already translated from the backend or translate them in the front-end?

nathangavin replied:
recoupled item value from the classes set

michalkleiner commented on plugins/PrivacyManager/vue/src/Compliance/Compliance.vue:
```diff
@@ -9,15 +9,74 @@
+  </p>
+
+  <SiteSelector
+    id="complianceDashboard"
```
Checked other uses of SiteSelector and they all have the id because they are also preceded by a <label> tag (being a form field of sorts), I reckon we should also add a label above the field, amend the id so it's clearer it's a site selector, e.g. complianceSite.

nathangavin replied:
added the label

michalkleiner commented on plugins/PrivacyManager/vue/src/Compliance/ComplianceTable.vue:
```diff
@@ -0,0 +1,57 @@
+          'non-compliant': item.value === 'non_compliant',
+          'unknown': item.value === 'unknown',
+        }">
+          <span :class="{
+              'icon-ok': item.value === 'compliant',
```
The `icon` class is in the stylesheet, but not used.. let's add it to maintain consistent icons width across the rows, otherwise each icon has slightly different width.

```suggestion
          <span :class="{
              'icon': true,
              'icon-ok': item.value === 'compliant',
```

nathangavin replied:
added extra class

michalkleiner commented on plugins/PrivacyManager/vue/src/Compliance/ComplianceTable.vue:
```diff
@@ -0,0 +1,57 @@
+        <td :class="{
+          'status': true,
+          'compliant': item.value === 'compliant',
+          'non-compliant': item.value === 'non_compliant',
```
If we returned `non-compliant` as the value for the requirement in the data (instead of the version with an underscore), we can directly use it as the CSS class.

nathangavin replied:
recoupled item value from the classes set

michalkleiner commented on plugins/PrivacyManager/vue/src/Compliance/ComplianceOverview.vue:
```diff
@@ -0,0 +1,101 @@
+    <Field
+      v-if="!state.loading"
+      uicontrol="checkbox"
+      name="enableFeature"
```
The name attribute needs to be unique for each compliance type. Ideally we would also provide a per-compliance type unique id, otherwise we have duplicates on the page.

nathangavin replied:
updated

michalkleiner commented on plugins/PrivacyManager/vue/src/Compliance/ComplianceOverview.vue:
```diff
@@ -0,0 +1,101 @@
+      :title="translate('PrivacyManager_ComplianceEnforceCheckboxIntro')"
+      :introduction="translate('PrivacyManager_ComplianceEnforceCheckboxTitle')"
+      :inline-help="translate('PrivacyManager_ComplianceEnforceCheckboxHelp')"
+      v-model="isComplianceModeEnabled"
```
Possibly rename so that it indicates better what it would do?
```suggestion
      v-model="shouldEnforceComplianceMode"
```

nathangavin replied:
renamed

michalkleiner reviewed (changes requested):
Overall looks ok as a demo with mock data.

Left a few minor suggestions and questions, mainly on the structure and naming. Nothing major that would require refactoring. Happy to jump on a call and resolve those together.

nathangavin committed:
minor renames and remove excess html

nathangavin committed:
decouple requirement value from onscreen values

nathangavin committed:
made name value on checkbox unique

nathangavin committed:
add pairing label to site selector

nathangavin commented:
@michalkleiner I've addressed your minor suggestions.

nathangavin committed:
Merge branch '5.x-dev' into dev-19376-UI

innocraft-automation committed:
Build vue files

nathangavin committed:
Merge branch '5.x-dev' into dev-19376-UI

nathangavin committed:
standardise variable names

michalkleiner committed:
Use camel case for compliance store data

michalkleiner committed:
Add missing component import

michalkleiner committed:
Build dist files

# Conflicts:
#	plugins/PrivacyManager/vue/dist/PrivacyManager.umd.js
#	plugins/PrivacyManager/vue/dist/PrivacyManager.umd.min.js

innocraft-automation committed:
Build vue files

michalkleiner committed:
Remove p from inside label

michalkleiner reviewed (dismissed):
Looks good now, thanks @nathangavin.
Let's wait for the tests to pass, then we can merge.

nathangavin committed:
update broken UI tests

michalkleiner committed:
Update expected UI test screenshot

# Conflicts:
#	plugins/PrivacyManager/tests/UI/expected-screenshots/PrivacyManager_compliance.png

michalkleiner committed:
Update UI test screenshot from CI
